Filename can be a nested path

sebdeckers opened the pull request:
*TL;DR* This patch allows setting of arbitrary filename to support uploads of directories. This matches current spec and behaviour of Chrome.

### Example

```js
for (const filename of ['foo/bar.txt', 'foo/baz.txt']) {
  form.append('uploads', createReadStream(filename), filename)
}
// Content-Disposition: form-data; name="uploads"; filename="foo/bar.txt"
// Content-Disposition: form-data; name="uploads"; filename="foo/baz.txt"
```

### Description

The [FormData `.append`](https://xhr.spec.whatwg.org/#dom-formdata-append) spec defers to the [File API](https://w3c.github.io/FileAPI/#dfn-name) spec which states:

> There are numerous file name variations and conventions used by different underlying OS file systems; this is merely the name of the file, without path information.

However, the [`HTMLInputElement.webkitdirectory `](https://wicg.github.io/entries-api/#dom-htmlinputelement-webkitdirectory) spec supports relative paths, and Chrome sets them as the `filename=...` in the form field.

This is also supported by [Multer](https://github.com/expressjs/multer), for example.

### Caveat Emptor

Breaking change for anyone manually specifying filename as a full path. Does not change behaviour on auto-detecting of filename from stream, http, etc.

### See also

- https://github.com/form-data/form-data/pull/203#issuecomment-218624023

sebdeckers commented:
😱 What's wrong with CI tests?! Worked fine for me. 🤔

sebdeckers commented:
The CodeFactor issues seem to be unrelated to this patch, AFAICT.

alexindigo commented:
Thanks for the PR @sebdeckers. Do you mind to add a test for the use case?
Also, how strong do you feel about using `filename` for the full path?
If we'd add something like `filepath` – it won't be breaking change.

Thanks.

PS. I'll check what's wrong with CodeFactor.

alexindigo commented:
I removed CodeFactor (it was weird and inconsistent) – it should be better now.

sebdeckers commented:
Ooh, that's a great idea! 💡 How did I not think of just adding a new property. 😅

Will refactor this PR and add testing.

sebdeckers commented:
All right. Took me a few tries to get Windows support working but this should do the trick. 😅 Appveyor is very useful.

alexindigo commented on lib/form_data.js:
```diff
@@ -211,18 +212,19 @@ FormData.prototype._getContentDisposition = function(value, options) {
 
   if (filename) {
-    contentDisposition = 'filename="' + path.basename(filename) + '"';
+    filename = options.filepath ? normalize(filename) : path.basename(filename);
```
How about using `path.normalize`?

sebdeckers replied:
Different purpose:

- `path.normalise` cleans up `..` and de-dupes path separators, but retains platform-specific path separators.
- `normalize-path` converts Windows `\` path separator to forward slash.

TBH I didn't check if the spec calls for forward slashes but seems consistent with all web-related paths.

alexindigo replied:
I see.

Probably we still want to clean up don't, `'./photos/toys/unicycle.jpg'` -> `'photos/toys/unicycle.jpg'`, as for win -> unix, we could do something like:
```
filename = filename.replace('\\', '/');
```

sebdeckers replied:
Okay!

alexindigo commented on lib/form_data.js:
```diff
@@ -211,18 +212,19 @@ FormData.prototype._getContentDisposition = function(value, options) {
   // fs- and request- streams have path property
   // formidable and the browser add a name property.
-  var filename = options.filename || value.name || value.path;
+  var filename = options.filepath || options.filename || value.name || value.path;
```
What do you think doing it in "one-stroke" fashion?

Something like:
```
var filename = options.filepath ? path.normalize(options.filepath) : path.basename(options.filename || value.name || value.path);
```

And add `basename` call to the streams.

The main reason is to handle options here and don't "leak" them further into the function.

sebdeckers replied:
Totally agree. Will clean this up.

alexindigo commented:
Thanks for the update. Looks good in general, just want to discuss implementation details.

sebdeckers commented:
(Sorry for the coveralls spam ... It seems to leave a comment on every rebase.)

Fixed as per review:
- Removed `normalize-path` dependency in favour of `.replace('\\', '/')`
- Refactored `filename` assignment to avoid mutating the variable.

Also fixed:
- `contentType` detection based on both `options.filepath` and `options.filename`

alexindigo commented:
Don't worry about coveralls :) With greenkeeper gone from this repo, it's all nice and quiet :)
And thank you for walking extra mile :)

alexindigo commented on lib/form_data.js:
```diff
@@ -209,20 +209,25 @@ FormData.prototype._multiPartHeader = function(field, value, options) {
-    filename = value.client._httpMessage.path;
+  if (typeof options.filepath === 'string') {
+    // custom filepath for relative paths
+    filename = path.normalize(options.filepath.replace(/\\/g, '/'));
```
Haha :) My bad. Sorry 👍

alexindigo commented on test/integration/test-custom-filename.js:
```diff
@@ -8,12 +8,14 @@ var assert       = common.assert;
 
 var knownFile = common.dir.fixture + '/unicycle.jpg';
 var unknownFile = common.dir.fixture + '/unknown_file_type';
+var relativeFile = path.relative('.', knownFile);
```
Maybe just try to hardcode stuff here, like `./photos/toys/unicycle.jpg`?

alexindigo commented on test/integration/test-custom-filename.js:
```diff
@@ -35,6 +37,10 @@ var server = http.createServer(function(req, res) {
     assert.strictEqual(files['custom_filename'].type, mime.lookup(knownFile), 'Expects original content-type');
 
+    assert('custom_filepath' in files);
+    assert.strictEqual(files['custom_filepath'].name, relativeFile.replace(/\\/g, '/'), 'Expects custom filename');
```
I remember struggles like that, with no local access to windows :)

You could try something like:

```
assert.strictEqual(files['custom_filepath'].name, relativeFile.replace(/\\/g, '/'), 'Expects custom filename:' + files['custom_filepath'].name + ' -vs- ' + relativeFile.replace(/\\/g, '/') + ' -from- ' + relativeFile + ' -end-');
```

:)

sebdeckers replied:
Hah this is too funny. 😂

Anyway I think it's because I did `path.resolve` after the backslash replace. Seems on Windows `path.resolve` expects to still see the native OS separator. Makes sense, I suppose.

Should be fixed now. Also cleaned up some path concatenation in `text/common.js`, just to be on the safe side.

alexindigo replied:
I see. Native nodejs methods do convert slashes for windows, so it should be fine.

alexindigo replied:
Oh, do you mind to change `'Expects custom filename'` to `'Expects custom filepath'`, it will be more clear which line is failed.

sebdeckers commented on test/integration/test-custom-filename.js:
```diff
@@ -8,12 +8,14 @@ var assert       = common.assert;
-var unknownFile = common.dir.fixture + '/unknown_file_type';
+var knownFile = path.join(common.dir.fixture, 'unicycle.jpg');
+var unknownFile = path.join(common.dir.fixture, 'unknown_file_type');
+var relativeFile = path.relative(path.join(__dirname, '..'), knownFile);
```
Oops this should have been relative to knownFile, not __dirname. One more patch incoming...

alexindigo replied:
Do you want to pass `rootDir` to the `common.dir.root`, so we stay more DRY here?

sebdeckers replied:
Changed this to be a relative path to `knownFile`. So it will have the OS-specific separator. This makes the test more meaningful than a hardcoded string. LMK if you prefer a change.

alexindigo commented on test/common.js:
```diff
@@ -10,9 +10,9 @@ var common = module.exports;
-  tmp: path.join(rootDir, '/test/tmp')
+  lib: path.join(rootDir, 'lib'),
+  fixture: path.join(rootDir, 'test', 'fixture'),
+  tmp: path.join(rootDir, 'test', 'tmp')
```
`path.join` updates all the slashes on windows.

sebdeckers replied:
Okay, will roll this back.

alexindigo commented:
🤞

alexindigo commented:
Looks like it's in good working state. I invited you as a Collaborator, you can merge when it's passes. Thank you.

sebdeckers commented:
@alexindigo Thanks! It's an honour. Appreciate your guidance & patience. 🤝

alexindigo commented:
Thank you. I'll publish it shortly.

alexindigo commented:
Published `v2.2.0`. Thanks again.
